### ISSUE TITLE
Error happenss when launching a subprocess using openmpi

### ISSUE THREAD

**Issue by cshanbo** [open-mpi/ompi#3158]
Hi openmpi team,

I am attempting run a [Theano](https://github.com/Theano/Theano)-based `Python` code for [neural machine translation](https://github.com/rsennrich/nematus) with openmpi and a [multi-node framework](https://github.com/cshanbo/platoon/tree/fix/multi-node) , but something weird happened. 
Any idea will be helpful and appreciated.

Background: The process of my codes is like:
1. starting a training process
2. starting a subprocess after some batches are trained (`Popen` or `exec_que` an external validation shell script). 

The weird things are,
1. Single-node Scenario
when I use `single-node` training, which only launch the training process on a single node (machine), without `mpirun`, an exception still encountered in `validation, step 2`.
The error is like:
```
[nmyjs_104_22:33214] PMIX ERROR: NOT-FOUND in file src/client/pmix_client.c at line 199
Training time 236.7153s
Number of samples: 4010
[nmyjs_104_22:33214] [[INVALID],INVALID] ORTE_ERROR_LOG: Unknown data type in file runtime/data_type_support/orte_dt_print_fns.c at line 161
[nmyjs_104_22:33214] *** Process received signal ***
[nmyjs_104_22:33214] Signal: Segmentation fault (11)
[nmyjs_104_22:33214] Signal code: Invalid permissions (2)
[nmyjs_104_22:33214] Failing at address: 0x2568700
[nmyjs_104_22:33214] [ 0] /lib64/libpthread.so.0[0x3da4a0f710]
[nmyjs_104_22:33214] [ 1] [0x2568700]
[nmyjs_104_22:33214] *** End of error message ***
```
I'm not sure why openmpi influenced the experiment in a `single node` scenario. (Although importing some `Theano` related module invokes the `fork` WARNING, I don't think that's related to this issue)

2. Multi-node scenario
Similar exception happens when using `mpirun -np 2 -H host1,host2 python xxx`.
The training process works fine util a subprocess, validation, is called. Similar error is invoked on two nodes, like:
```
*** glibc detected *** python: double free or corruption (!prev): 0x0000000002ba9f10 ***
[nmyjs_104_22:01210] *** Process received signal ***
[nmyjs_104_22:01210] Signal: Segmentation fault (11)
[nmyjs_104_22:01210] Signal code: Invalid permissions (2)
[nmyjs_104_22:01210] Failing at address: 0x29405e0
[nmyjs_104_22:01210] [ 0] /lib64/libpthread.so.0[0x3da4a0f710]
[nmyjs_104_22:01210] [ 1] [0x29405e0][nmyjs_104_22:01210] *** End of error message ***

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator <Unknown>
with errorcode 0.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.

[nmyjs_104_22:51831] OPAL ERROR: Timeout in file base/pmix_base_fns.c at line 193
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_dpm_dyn_init() failed
  --> Returned "Timeout" (-15) instead of "Success" (0)
--------------------------------------------------------------------------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator <Unknown>
with errorcode 0.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

### example code

When the `req` is `valid_err` (examples of what did `req` do can be found [here](https://github.com/cshanbo/platoon/blob/fix/multi-node/platoon/channel/controller.py#L206) ), invoking a subprocess to validate.
```
        elif 'valid_err' in req:
            valid_err = req['valid_err']
            print("valid:", valid_err)
            self.history_errs.append(valid_err)
            harr = numpy.array(self.history_errs)

            if self.external_validation_script:
                print("Calling external validation script")
                cmd = [self.external_validation_script, str(self.valid_gpu), str('%s/model.npz'%self.model_dir), str(self.last_save_index), str(self.valid_datadir)]
                self.exec_que.put(cmd)
            if valid_err <= harr.min():
                self.bad_counter = 0
                control_response = 'best'
                print("Best error valid:", valid_err)
            elif (len(self.history_errs) > self.patience and valid_err >= harr[:-self.patience].min()):
                self.bad_counter += 1
                print("Bad Counter add one, now:", self.bad_counter)
```

If anybody has any idea, please help. Thank you so very much.

**Comment by rhc54** [comment]
From you first scenario, it looks like your validation script is invoking an application that calls MPI_Init. If you do that directly (i.e., without using mpirun), then it runs as a "singleton". The error you received might indicate that the app was built against one version of OMPI, but your current LD_LIBRARY_PATH and PATH are pointing at another.

**Comment by cshanbo** [comment]
Hi @rhc54 
Thank you for you help!

> The error you received might indicate that the app was built against one version of OMPI, but your current LD_LIBRARY_PATH and PATH are pointing at another

I don't think I confused OMPI. I installed open-mpi in the default path, which is `/usr/loca/`, and I set the `PATH` and `LD_LIBRARY_PATH` to `/usr/local/bin`, `/usr/local/lib`, separately to make sure the environmental variables are correct.

Basically, what my validation script is to call a python code like [this file](https://github.com/rsennrich/nematus/blob/master/nematus/translate.py) , which is, loading a model, then test the model on a small validation dataset. 

Could you please look at [translate.py](https://github.com/rsennrich/nematus/blob/master/nematus/translate.py) briefly to see if there's anything that breaks mpi? 

btw, the validate script is as simple as:
```
THEANO_FLAGS=mode=FAST_RUN,floatX=float32,device=$device,on_unused_input=warn python \
  $bin/translate.py \
       -m $prefix.$iter.npz \
       -i $dev \
       -o $translate_outputfile \
       -k 12 -n -p 1
```

Thank you so much.

**Comment by cshanbo** [comment]
let me use an example to explain another weird thing:
```
configuration:
    valid_freq=2000,
    batch_size=10,
```
Training process will invoke `validation script` every 2000 batches.
When invoking the validation for the first time, the validation script runs fine (training after the first 2000 batches).
The validation process will raise the error mentioned above, starts from the second validation (4000 batches in this example).

**Comment by rhc54** [comment]
I'll go back to what I suggested on the other issue you filed: start by running simple MPI examples to ensure your installation is correct (we provide some in the OMPI examples directory). Then you can add a simple Python-MPI example (mpi4py provides a number of them). If all those work, then the likely problem is in your program and not the OMPI code.

**Comment by ggouaillardet** [comment]
@cshanbo what is your subprocess and how is it launched ?
is your subprocess a MPI application ?
is it launched with ```MPI_Comm_spawn()``` or ```fork();exec(...)```

if your MPI app bluntly fork&exec an other MPI app, then i strongly suspect the forked app will get very confused because it inherits the environment from the parent MPI app (e.g. same MPI rank).

**Comment by cshanbo** [comment]
@ggouaillardet I launched my python code with mpirun. And the python code will launch another script by `exec_que` or `Popen`. That's a `fork` indeed, so I think this might be the problem, but not sure how to fix it. Is there any solution to keep the `Popen` and not break the `mpi`? 

Thank you

**Comment by rhc54** [comment]
Ah, that is indeed not going to work. You cannot have a job launched via mpirun start another job using mpirun. If an MPI job wants to start another MPI job, then it is supposed to use MPI_Comm_spawn to do it.

**Comment by ggouaillardet** [comment]
@rhc54 would it be enough to remove all ```OMPI_*``` and ```PMIX_*``` from the environment before ```exec()``` the sub program ?
we would end up with basically two independent MPI apps (one or both being singleton fwiw)

**Comment by cshanbo** [comment]
Yes, if it is the 2nd scenario, I can understand that the mpi exception. But in my first scenario, i didn't use mpirun, but also raises the exception. This is weird. 

Best regards
Shanbo

**Comment by ggouaillardet** [comment]
@cshanbo if you do not use ```mpirun``` but your app calls ```MPI_Init()```, then you are in singleton mode.
```orted``` is fork&exec'ed under the hood, and from a high level point of view, this is pretty much equivalent to using ```mpirun -np 1 ...``` in the first place.

**Comment by rhc54** [comment]
@ggouaillardet People have tried your proposed approach, with mixed results. The problem is that you cannot tell which OMPI params should be kept, and which should be removed. If the user intends, for example, that an MCA param set in the environment actually govern all their MPI applications, then it won't work.

It also results in a program that isn't portable across MPI implementations, but perhaps that isn't a consideration here.

So bottom line is that @cshanbo can certainly try that approach and see if it works. If it does and there is no concern over portability, then it might be "good enough".

**Comment by cshanbo** [comment]
Hi all. I tried the approach that @ggouaillardet proposed, and that works!
Thank you for your help.
